### PR TITLE
Make QuadMesh arguments with defaults keyword_only

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

This is in preparation of changing the signature from 

`__init__(self, meshWidth, meshHeight, coordinates, antialiased=True, shading='flat', **kwargs)`

to 

`__init__(self, coordinates, *, antialiased=True, shading='flat', **kwargs)`

See also: https://github.com/matplotlib/matplotlib/pull/18472#issuecomment-841735177

If we want to make this according to deprecation rules this is a multi-step process:

- Make everything after `coordinates` keyword-only. (This PR).
- When that deprecation expires (in 3.7), change to
  `__init__(self, *args, antialiased=True, shading='flat', **kwargs)` and extract `meshWidth, meshHeight, coordinates` from `args, kwargs`. At the same time, deprecate passing `meshWidth, meshHeight, 1d_coordinates` in favor of passing 2D coordinates only.
- When that deprecation expires (in 3.9), change back to `__init__(self, coordinates, *, antialiased=True, shading='flat', **kwargs)`

The in between `*args` step is essential for a smooth transition phase allowing both signatures `width, height, 1d_coordinates` and `2d_coordinates`.

Possible shortcut: Skip the first step and directly go to the `*args` signature. This would be a breaking change, but only for people, who pass `antialiased` as positional argument.

### CODE REVIEW

**Comment by greglucas** [comment]
Nice, I think you are much more lenient than I was going to be. I had this following diff ready to go, which would be your second choice. I'd still vote for skipping the keyword-only deprecation period as I can't imagine many people relying on that "feature".

```diff
diff --git a/lib/matplotlib/collections.py b/lib/matplotlib/collections.py
index d02607cc5..bcce49440 100644
--- a/lib/matplotlib/collections.py
+++ b/lib/matplotlib/collections.py
@@ -1977,34 +1977,38 @@ class QuadMesh(Collection):
     (*meshWidth* * *meshHeight*) quadrilaterals in the mesh.  The mesh
     need not be regular and the polygons need not be convex.
 
-    A quadrilateral mesh is represented by a (2 x ((*meshWidth* + 1) *
-    (*meshHeight* + 1))) numpy array *coordinates*, where each row is
-    the *x* and *y* coordinates of one of the vertices.  To define the
+    A quadrilateral mesh is represented by a ((*meshWidth* + 1),
+    (*meshHeight* + 1), 2) numpy array *coordinates*, where the final dimension
+    is the *x* and *y* coordinates of one of the vertices.  To define the
     function that maps from a data point to its corresponding color,
-    use the :meth:`set_cmap` method.  Each of these arrays is indexed in
-    row-major order by the mesh coordinates of the vertex (or the mesh
-    coordinates of the lower left vertex, in the case of the colors).
-
-    For example, the first entry in *coordinates* is the coordinates of the
-    vertex at mesh coordinates (0, 0), then the one at (0, 1), then at (0, 2)
-    .. (0, meshWidth), (1, 0), (1, 1), and so on.
+    use the :meth:`set_cmap` method.
 
     *shading* may be 'flat', or 'gouraud'
     """
-    def __init__(self, meshWidth, meshHeight, coordinates,
+    def __init__(self, coordinates, *args,
                  antialiased=True, shading='flat', **kwargs):
         super().__init__(**kwargs)
-        self._meshWidth = meshWidth
-        self._meshHeight = meshHeight
-        # By converting to floats now, we can avoid that on every draw.
-        self._coordinates = np.asarray(coordinates, float).reshape(
-            (meshHeight + 1, meshWidth + 1, 2))
+        if len(args):
+            _api.warn_deprecated("3.5", alternative=(
+                "The QuadMesh class now takes a 3-dimensional *coordinates* "
+                "array as input of shape (nx, ny, 2) that defines the shape "
+                "and size of the mesh."))
+            self._meshWidth = coordinates
+            self._meshHeight = args[0]
+            # By converting to floats now, we can avoid that on every draw.
+            self._coordinates = np.asarray(args[1], float).reshape(
+                (self._meshHeight + 1, self._meshWidth + 1, 2))
+        else:
+            # This will become standard after the deprecation
+            self._coordiantes = np.asarray(coordinates, float)
+            self._meshWidth = self._coordinates.shape[0] - 1
+            self._meshHeight = self._coordinates.shape[1] - 1
+
         self._antialiased = antialiased
         self._shading = shading
 
         self._bbox = transforms.Bbox.unit()
-        self._bbox.update_from_data_xy(coordinates.reshape(
-            ((meshWidth + 1) * (meshHeight + 1), 2)))
+        self._bbox.update_from_data_xy(self._coordinates.reshape(-1, 2))
 
     def get_paths(self):
         if self._paths is None:
```

**Comment by jklymak** [comment]
I'd personally vote for the simpler solution.  I expect vanishingly few users use `QuadMesh` directly, and I doubt any developers who use it are going to be using the kwargs positionally.

**Comment by anntzer** [comment]
I think this can be done by manually writing the deprecation instead?  Something like
```python
def __init__(self, *args, **kwargs):
    old_sig = inspect.signature(lambda self, meshWidth, ...)
    new_sig = inspect.signature(lambda self, coordinates, ...)
    try:
        bound = new_sig.bind(*args, **kwargs)
        # extract arguments, do the actual constructor work
        ...
    except TypeError as exc:
        try:
            bound = old_sig.bind(*args, **kwargs)
            # user passed in args using the old call convention
            warn_deprecated(...)
            # extract arguments, do the actual constructor work
            ...
        except TypeError:
            raise exc  # probably the new_sig binding error makes more sense

# so that docs and introspection display the new signature
__init__.__signature__ = inspect.signature(
    lambda self, coordinates, *, antialiased=True, ...)
```
taking advantage of the fact that we can distinguish an old-style call from a new-style call just from the passed in args?  (even ignoring arg types, I don't think(?) a call can ever be correct for both signatures)

**Comment by timhoffm** [comment]
A major part of this PR does not come from the deprecation logic itself. It's rather adapting our internal calls to the new API and  tests that the old API still works and produces the same result as the new API.

File: lib/matplotlib/collections.py [hunk 1]
```diff
@@ -1966,6 +1968,34 @@ class QuadMesh(Collection):
...
+      :            /                   /
+      :           /                 /
+      :          /               /
+      :      (m, n) -------- (m, n+1)
+
+    The mesh need not be regular and the polygons need not be convex.
+
+    Parameters
+    ----------
+    coordimates : (M, N, 2) array-like
```
Comment (anntzer) [inline lib/matplotlib/collections.py:1984]: typo: mates -> nates

File: lib/matplotlib/collections.py [hunk 2]
```diff
@@ -1989,23 +2019,75 @@ class QuadMesh(Collection):
...
+            lambda meshWidth, meshHeight, coordinates,
+                   antialiased=True, shading='flat', **kwargs: None)
+        new_sig = inspect.signature(
+            lambda coordinates, *,
+                   antialiased=True, shading='flat', **kwargs: None)
+        try:
+            bound = new_sig.bind(*args, **kwargs)
+        except TypeError as exc:
+            try:
+                print(args, kwargs)
```
Comment (anntzer) [inline lib/matplotlib/collections.py:2036]: remove the print

File: lib/matplotlib/collections.py [hunk 3]
```diff
@@ -1989,23 +2019,75 @@ class QuadMesh(Collection):
...
+                bound = old_sig.bind(*args, **kwargs)
+            except TypeError:
+                raise exc  # the new_sig binding error makes more sense
+                # user passed in args using the old call convention
+            else:
+                bound.apply_defaults()
```
Comment (anntzer) [inline lib/matplotlib/collections.py:2042]: Sorry for now realizing this earlier, but a simpler solution would be to do something like
```python
try:
    # The following raises a TypeError iif the args don't match.
    w, h, c, a, s, k = (
        lambda meshWidth, meshHeight, coordinates, antialiased=True, shading=False, **kwargs:
        (meshWidth, meshHeight, coordinates, antialiased, shading, kwargs))(*args, **kwargs)
except TypeError as exc:
    # If the following raises a TypeError (i.e. args don't match), just let it propagate.
    c, a, s = (
        lambda coordinates, antialiased=True, shading=False, **kwargs:
        (coordinates, antialiased, shading, kwargs))(*args, **kwargs)
else:  # The old signature matched.
    warn_deprecated(...)
    c = np.asarray(c, float).reshape((h + 1, w + 1, 2))
```
avoiding the need to `apply_defaults()` and to extract args from `bound`.

File: lib/matplotlib/collections.py [hunk 4]
```diff
@@ -1989,22 +2019,60 @@ class QuadMesh(Collection):
...
+
         super().__init__(**kwargs)
-        self._meshWidth = meshWidth
-        self._meshHeight = meshHeight
-        # By converting to floats now, we can avoid that on every draw.
-        self._coordinates = np.asarray(coordinates, float).reshape(
-            (meshHeight + 1, meshWidth + 1, 2))
+        self._coordinates = coords
+        shape = self._coordinates.shape
+        if (self._coordinates.ndim != 3 or shape[-1] != 2):
```
Comment (anntzer) [inline lib/matplotlib/collections.py:2059]: that's _api.check_shape?

File: lib/matplotlib/collections.py [hunk 5]
```diff
@@ -1989,22 +2019,55 @@ class QuadMesh(Collection):
...
+    def __init__(self, *args, **kwargs):
+        # signature deprecation since="3.5": Change to new signature after the
+        # deprecation has expired. Also remove setting __init__.__signature__,
+        # and remove the Notes from the docstring.
+        #
+        # We use lambdas to parse *args, **kwargs through the respective old
+        # and new signatures.
+        try:
+            # Old signature:
+            # The following raises a TypeError iif the args don't match.
```
Comment (jklymak) [inline lib/matplotlib/collections.py:2032]: Maybe explain a little bit what this does?
```suggestion
            # The following raises a TypeError iif the args don't match.  The lambda tries to 
            # assign the *args, and **kwargs to the old-style signature and fails if the 
            # new style signature is being used.  
```

Comment (timhoffm) [reply lib/matplotlib/collections.py:2032]: We have the comment above
```
       # We use lambdas to parse *args, **kwargs through the respective old
        # and new signatures.
```
which I think is good enough given that this is just a temporary solution during the deprecatoin.

Comment (jklymak) [reply lib/matplotlib/collections.py:2032]: Ooops sorry missed that...

**Review by jklymak — APPROVED** [review]

File: doc/api/next_api_changes/deprecations/20237-TH.rst [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+is deprecated and replaced by the new signature ::
+
+    def __init__(coordinates, *, antialiased=True, shading=False, **kwargs)
+
+In particular:
+
+- *coordinates* must now be a (M, N, 2) array-like. Previously, the grid
+  shape was separately specified as (*meshHeight* + 1, *meshWidth* + 1) and
+  *coordnates* could be an array-like of any shape with M * N * 2 elements.
```
Comment (anntzer) [inline doc/api/next_api_changes/deprecations/20237-TH.rst:16]: typo: coordnates

**Review by anntzer — APPROVED** [review]
can selfmerge after fixing typos

File: lib/matplotlib/collections.py [hunk 7]
```diff
@@ -1966,6 +1968,34 @@ class QuadMesh(Collection):
     """
     Class for the efficient drawing of a quadrilateral mesh.
 
+    A quadrilateral mesh consists of a (N, M) grid of vertices, that define
```
Comment (anntzer) [inline lib/matplotlib/collections.py:1971]: (M, N)
and below
(M-1)*(N-1)

(to be consistent with the indices later)

File: lib/matplotlib/collections.py [hunk 8]
```diff
@@ -1966,6 +1968,34 @@ class QuadMesh(Collection):
...
+      :           /                 /
+      :          /               /
+      :      (m, n) -------- (m, n+1)
+
+    The mesh need not be regular and the polygons need not be convex.
+
+    Parameters
+    ----------
+    coordinates : (M, N, 2) array-like
+        The vertices. ``coordinates[m, n]`` specified the coordinates (x, y)
```
Comment (anntzer) [inline lib/matplotlib/collections.py:1985]: specifies the (x, y) coordinates of...

File: lib/matplotlib/collections.py [hunk 9]
```diff
@@ -1989,22 +2019,55 @@ class QuadMesh(Collection):
...
+            # just let it propagate.
+            coords, antialiased, shading, kwargs = (
+                lambda coordinates, antialiased=True, shading=False, **kwargs:
+                (coordinates, antialiased, shading, kwargs))(*args, **kwargs)
+            coords = np.asarray(coords, np.float64)
+        else:  # The old signature matched.
+            _api.warn_deprecated(
+                "3.5",
+                message="This usage of Quadmesh is deprecated: Parameters "
+                        "meshWidth and meshHights will be removed; "
```
Comment (anntzer) [inline lib/matplotlib/collections.py:2050]: typo: Hights

File: lib/matplotlib/collections.py (hunk 7 above)
Comment (jklymak) [reply lib/matplotlib/collections.py:1971]: All these PRs about quad mesh have been a bit confusing with respect to the indices convention.  I'd propose the convention is that MxN is the number of cells in the quadmesh and the coordinate are either M+1, N+1 for flat or M,N for gouraud?

Comment (anntzer) [reply lib/matplotlib/collections.py:1971]: I don't have a strong opinion wrt. +/-1 here.

**Pull request merged**